### PR TITLE
Added `databricks_pipelines` data source

### PR DESCRIPTION
**Pull request by boydZD**
## Changes
Adding data provider for DLT (pipelines) - [Issue-1926](https://github.com/databricks/terraform-provider-databricks/issues/1926)

## Tests
In addition to the test below, local tests using the developed data provider were conducted against an internal Databricks workspace.

- [x] `make test` run locally
- [x] relevant change in `docs/` folder
- [x] covered with integration tests in `internal/acceptance`
- [ ] relevant acceptance tests are passing
- [ ] using Go SDK

### CODE REVIEW

**Review by alexott — CHANGES_REQUESTED** [review]
please use `common.WorkspaceData` function instead [src](https://github.com/databricks/terraform-provider-databricks/blob/master/common/resource.go#L258), and try to use corresponding Go SDK function

**Comment by boydZD** [comment]
> 

updated to use common.WorkspaceData

File: docs/data-sources/pipeline.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+---
+subcategory: "Compute"
+---
+# databricks_pipeline Data Source
+
+-> **Note** If you have a fully automated setup with workspaces created by [databricks_mws_workspaces](../resources/mws_workspaces.md) or [azurerm_databricks_workspace](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/databricks_workspace), please make sure to add [depends_on attribute](../index.md#data-resources-and-authentication-is-not-configured-errors) in order to prevent _authentication is not configured for provider_ errors.
+
+Retrieves a name-to-id map of all [Delta Live Tables](https://docs.databricks.com/data-engineering/delta-live-tables/index.html) (DLT) pipelines deployed in a workspace, or those matching the provided search term.
```
Comment (nfx) [inline docs/data-sources/pipeline.md:8]: if it retrieves all pipelines, then data source should be plural - `databricks_pipelines`

**Review by nfx — CHANGES_REQUESTED** [review]

File: docs/data-sources/pipeline.md (hunk 1 above)
Comment (boydZD) [reply docs/data-sources/pipeline.md]: data_source and docs updated to reflect this

File: README.md [hunk 2]
```diff
@@ -60,6 +60,7 @@
 | [databricks_obo_token](docs/resources/obo_token.md)
 | [databricks_permissions](docs/resources/permissions.md)
 | [databricks_pipeline](docs/resources/pipeline.md)
+| [databricks_pipeline](docs/data-sources/pipeline.md) data
```
Comment (alexott) [inline README.md:63]: rename here as well

File: pipelines/data_pipelines.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+		Name string            `json:"pipeline_name,omitempty"`
+		Ids  map[string]string `json:"ids,omitempty" tf:"computed"`
+	}
+	return common.WorkspaceData(func(ctx context.Context, data *pipelineData, w *databricks.WorkspaceClient) error {
+		errorMessage := fmt.Errorf("no pipelines found")
+		pipelineSearch := pipelines.ListPipelines{MaxResults: 100}
+
+		if data.Name != "" {
+			searchPattern := fmt.Sprintf("name LIKE '%s'", data.Name)
+			errorMessage = fmt.Errorf("there is no pipeline with name LIKE '%s'; you need to specify `pipeline_name` as an `exact_name` or with percent wildcards", data.Name)
```
Comment (alexott) [inline pipelines/data_pipelines.go:24]: replace 

```
`exact_name` 
```

with something like "full name"?

**Review by alexott — CHANGES_REQUESTED** [review]
overall look good, just need to fix names in README

File: pipelines/data_pipelines.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+			pipelineSearch = pipelines.ListPipelines{Filter: searchPattern, MaxResults: 100}
+		}
+
+		pipelines, err := w.Pipelines.ListPipelinesAll(context.Background(), pipelineSearch)
+		if err != nil {
+			return err
+		} else if len(pipelines) == 0 {
+			return errorMessage
```
[the remark this answers is not in the document]
Comment (boydZD) [reply pipelines/data_pipelines.go:32]: here I was thinking to provide a more informative output in the event there are no pipelines defined or matching the search criteria provided via `pipeline_name`

File: pipelines/data_pipelines.go (hunk 3 above)
Comment (boydZD) [reply pipelines/data_pipelines.go]: adjusted text from "as an `exact_name`" to "as the full pipeline name"

File: pipelines/data_pipelines.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+		errorMessage := fmt.Errorf("no pipelines found")
+		pipelineSearch := pipelines.ListPipelines{MaxResults: 100}
+
+		if data.Name != "" {
+			searchPattern := fmt.Sprintf("name LIKE '%s'", data.Name)
+			errorMessage = fmt.Errorf("there is no pipeline with name LIKE '%s'; you need to specify `pipeline_name` as an `exact_name` or with percent wildcards", data.Name)
+			pipelineSearch = pipelines.ListPipelines{Filter: searchPattern, MaxResults: 100}
+		}
+
+		pipelines, err := w.Pipelines.ListPipelinesAll(context.Background(), pipelineSearch)
```
[the remark this answers is not in the document]
Comment (boydZD) [reply pipelines/data_pipelines.go]: yes, corrected.

File: pipelines/data_pipelines.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mgyucht) [reply pipelines/data_pipelines.go:32]: Is there no consistent pattern today? @nfx we should discuss whether we should have some consistency across all data sources w.r.t. how we deal with the absence of resources.

[the remark this answers is not in the document]
Comment (mgyucht) [reply pipelines/data_pipelines.go:32]: OK I just took a look at a handful of data source implementations (clusters, jobs, SQL warehouses) and all of them return empty list if no entries are defined. I think we can use that as a safe default, and users can potentially use [postconditions](https://developer.hashicorp.com/terraform/language/expressions/custom-conditions#preconditions-and-postconditions) if need be to assert that there is a pipeline defined. Later on, we can always add a parameter if we feel it is necessary.

[the remark this answers is not in the document]
Comment (alexott) [reply pipelines/data_pipelines.go:32]: I think that it makes sense to document behavior in the `CONTRIBUTING.md`:

- when you implement a data source that may return multiple results (clusters/jobs/...), then we need to return an empty list if nothing is found
- when you implement a data source that returns a single result (cluster/job/...) then we need to throw an error when no results or more than one result

File: internal/acceptance/data_pipelines_test.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+
+	"testing"
+)
+
+func TestAccDataSourcePipelines(t *testing.T) {
+	accountLevel(t, step{
+		Template: `
+		data "databricks_pipelines" "this" {
```
Comment (nfx) [inline internal/acceptance/data_pipelines_test.go:14]: can you create a pipeline - there might be a situation, where test workspace has no pipelines at all. e.g. acceptance tests must be self-contained.

File: pipelines/data_pipelines.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+		errorMessage := fmt.Errorf("no pipelines found")
+		pipelineSearch := pipelines.ListPipelines{MaxResults: 100}
+
+		if data.Name != "" {
+			searchPattern := fmt.Sprintf("name LIKE '%s'", data.Name)
+			errorMessage = fmt.Errorf("there is no pipeline with name LIKE '%s'; you need to specify `pipeline_name` as the full pipeline name or with percent wildcards", data.Name)
+			pipelineSearch = pipelines.ListPipelines{Filter: searchPattern, MaxResults: 100}
+		}
+
+		pipelines, err := w.Pipelines.ListPipelinesAll(ctx, pipelineSearch)
```
Comment (nfx) [inline pipelines/data_pipelines.go:28]: please use https://pkg.go.dev/github.com/databricks/databricks-sdk-go@v0.7.0/service/pipelines#PipelinesAPI.PipelineStateInfoNameToPipelineIdMap, which also performs name de-duplication.

**Review by nfx — CHANGES_REQUESTED** [review]

File: pipelines/data_pipelines.go (hunk 7 above)
Comment (boydZD) [reply pipelines/data_pipelines.go:28]: > please use https://pkg.go.dev/github.com/databricks/databricks-sdk-go@v0.7.0/service/pipelines#PipelinesAPI.PipelineStateInfoNameToPipelineIdMap, which also performs name de-duplication.

#nfx Pipeline names are not unique. See "test" pipelines in screenshot. In this scenario, should we throw an error when searching by pipeline name "test" or return all pipelines that match that search criteria?
Screenshot 2023-04-12 at 10 13 45 AM

Comment (alexott) [reply pipelines/data_pipelines.go:28]: imho, for `databricks_pipelines` we should return all of them.

Comment (boydZD) [reply pipelines/data_pipelines.go:28]: that's my thought as well. If that's the case then we should return id-to-name map (ID as key) from this data provider to allow for duplicate names.

File: internal/acceptance/data_pipelines_test.go (hunk 6 above)
Comment (boydZD) [reply internal/acceptance/data_pipelines_test.go:92]: acceptance test updated to include pipeline creation

File: pipelines/data_pipelines.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/databricks/databricks-sdk-go"
+	"github.com/databricks/databricks-sdk-go/service/pipelines"
+	"github.com/databricks/terraform-provider-databricks/common"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func DataSourcePipelines() *schema.Resource {
+	type pipelineData struct {
+		Name string            `json:"pipeline_name,omitempty"`
+		Ids  map[string]string `json:"ids,omitempty" tf:"computed"`
```
Comment (nfx) [inline pipelines/data_pipelines.go:16]:
```suggestion
		Ids  []string `json:"ids,omitempty" tf:"computed,slice_set"`
```

we don't have a pattern of "id-to-name" anywhere. please use the set and sort it, before returning - it'll make unit tests predictable.

1. name to id: https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/mws_credentials#ids
2. set: https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/clusters#ids
3. name to id: https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/jobs#ids
4. set https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/sql_warehouses#ids
5. name to id https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/mws_workspaces#ids
6. set https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/service_principals#application_ids
7. set https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/dbfs_file_paths#path_list
8. set https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/catalogs#ids
9. set https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/schemas#ids
10. set https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/shares#shares
11. set https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/tables#ids
12. set https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/views#ids
13. set https://registry.terraform.io/providers/databricks/databricks/latest/docs/data-sources/notebook_paths#notebook_path_list

**Review by nfx — CHANGES_REQUESTED** [review]

File: pipelines/data_pipelines.go (hunk 8 above)
Comment (boydZD) [reply pipelines/data_pipelines.go]: Refactored to return a list of ids as discussed.

**Review by nfx — APPROVED** [review]

**Pull request merged**